### PR TITLE
Add PipelineOverview component and show pipeline list in sidebar

### PR DESCRIPTION
**Pull request by sahil143**
ODC story: https://jira.coreos.com/browse/ODC-1799

Screenshot from 2019-10-23 19-15-00

### CODE REVIEW

File: frontend/packages/dev-console/src/components/pipelines/pipeline-overview/PipelineOverview.tsx [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+import { pipelineRunFilterReducer } from '../../../utils/pipeline-filter-reducer';
+import TriggerLastRunButton from './TriggerLastRunButton';
+
+const noOfPipelineRunsToshow = 3;
+
+type PipelineRunItemProps = {
+  pipelineRun: K8sResourceKind;
+};
+
+export const PipelineRunItem: React.FC<PipelineRunItemProps> = ({ pipelineRun }) => {
```
Comment (rohitkrai03) [inline frontend/packages/dev-console/src/components/pipelines/pipeline-overview/PipelineOverview.tsx]: Move this component out into its separate file.

File: frontend/packages/dev-console/src/components/pipelines/pipeline-overview/PipelineOverview.tsx [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+    pipelineRuns,
+  },
+}) => {
+  const {
+    metadata: { name, namespace },
+  } = pipeline;
+  return (
+    <React.Fragment>
+      <SidebarSectionHeading text="Pipeline Run" style={{ display: 'inline-block' }} />
+      {_.size(pipelineRuns) > noOfPipelineRunsToshow && (
```
Comment (rohitkrai03) [inline frontend/packages/dev-console/src/components/pipelines/pipeline-overview/PipelineOverview.tsx]: Why not just use `pipelineRuns.length`? Its going to be a bit faster than lodash function. `_.size()` would be more helpful when the collection could be both an array or an object. But in this case it's always going to be an array. Plus the implementation uses `length` property of an array internally as well.

File: frontend/packages/dev-console/src/components/pipelines/pipeline-overview/PipelineOverview.tsx [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+        <Link
+          className="sidebar__section-view-all"
+          to={`${resourcePath(referenceForModel(PipelineModel), name, namespace)}/Runs`}
+        >
+          {`View All (${_.size(pipelineRuns)})`}
```
Comment (rohitkrai03) [inline frontend/packages/dev-console/src/components/pipelines/pipeline-overview/PipelineOverview.tsx]: Same as above.

File: frontend/public/components/overview/resource-overview-page.tsx [hunk 4]
```diff
@@ -9,19 +9,38 @@ import { NetworkingOverview } from './networking-overview';
...
+}) => {
+  const { buildConfigs, routes, services, pods, obj } = item;
+  const pluginComponents = getPluginTabSectionResource(item);
+  return (
+    <div className="overview__sidebar-pane-body">
+      <PodsOverview pods={pods} obj={obj} />
+      <BuildOverview buildConfigs={buildConfigs} />
+      {pluginComponents.map((Component, index) => (
+        <Component key={index} item={item} />
+      ))}
```
Comment (andrewballantyne) [inline frontend/public/components/overview/resource-overview-page.tsx]: This appears to be constantly mounted and unmounted. I am unsure why - even when using a static key it still happens. I'll continuing investigating.

File: frontend/packages/dev-console/src/components/pipelines/pipeline-overview/PipelineOverview.tsx [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+    pipeline: [pipeline],
+    pipelineRuns,
+  },
+}) => {
+  const {
+    metadata: { name, namespace },
+  } = pipeline;
+  return (
+    <React.Fragment>
+      <SidebarSectionHeading text="Pipeline Run" style={{ display: 'inline-block' }} />
```
[the remark this answers is not in the document]
Comment (andrewballantyne) [reply frontend/packages/dev-console/src/components/pipelines/pipeline-overview/PipelineOverview.tsx]: Would recommend using the model here instead of a hardcoded `Pipeline Run` text. First, we should avoid disjointed text that should be connected, and two the other sections are plural -- so we should be plural as well.

```suggestion
      <SidebarSectionHeading text={PipelineRunModel.plural} style={{ display: 'inline-block' }} />
```

File: frontend/packages/dev-console/src/components/pipelines/pipeline-overview/PipelineOverview.tsx [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+                inline
+                kind={referenceForModel(PipelineModel)}
+                name={name}
+                namespace={namespace}
+              />
+            </GridItem>
+            <GridItem span={4} className="text-right">
+              <TriggerLastRunButton pipeline={pipeline} />
+            </GridItem>
+          </Grid>
```
Comment (andrewballantyne) [inline frontend/packages/dev-console/src/components/pipelines/pipeline-overview/PipelineOverview.tsx]: This Grid is unbalanced - There is a notable gap of whitespace between the end of your first GridItem and the TriggerButton. This space could be used for the title when there are longer pipeline names.

Would recommend using flex to close the gap. A minor set of padding should be all you need to buffer the title from the button, otherwise the title should be allowed to expand and use the space as it needs.

Look at the way the Builds section is doing the title and you'll see the gap is more effectively used.

File: frontend/packages/dev-console/src/components/pipelines/pipeline-overview/PipelineOverview.tsx [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+import { PipelineRunModel, PipelineModel } from '@console/dev-console/src/models';
+import { Link } from 'react-router-dom';
+import { fromNow } from '@console/internal/components/utils/datetime';
+import { pipelineRunFilterReducer } from '../../../utils/pipeline-filter-reducer';
+import TriggerLastRunButton from './TriggerLastRunButton';
+
+const noOfPipelineRunsToshow = 3;
+
+type PipelineRunItemProps = {
+  pipelineRun: K8sResourceKind;
```
Comment (andrewballantyne) [inline frontend/packages/dev-console/src/components/pipelines/pipeline-overview/PipelineOverview.tsx]: We can use the `PipelineRun` type here, since it's in the `dev-console` package.

File: frontend/public/components/overview/resource-overview-page.tsx [hunk 8]
```diff
@@ -9,19 +9,38 @@ import { NetworkingOverview } from './networking-overview';
...
-  </div>
+const getResourceTabSectionComp = (t) => (props) => (
+  <AsyncComponent {...props} loader={t.properties.loader} />
 );
 
+const getPluginTabSectionResource = (item) => {
+  const sections = plugins.registry
+    .getOverviewTabSections()
+    .filter((section) => item[section.properties.key]);
+  return sections.map((section) => getResourceTabSectionComp(section));
```
Comment (andrewballantyne) [inline frontend/public/components/overview/resource-overview-page.tsx]: You could just chain it all together:
```
  return plugins.registry
    .getOverviewTabSections()
    .filter((section) => item[section.properties.key])
    .map((section) => getResourceTabSectionComp(section));
```

Comment (andrewballantyne) [inline frontend/public/components/overview/resource-overview-page.tsx]: You could also return a key based off of the [type](https://github.com/openshift/console/pull/3056/files#diff-34d17997d30d17137995c2847e41e305R37) that you could use in your `pluginComponents.map` call [below](https://github.com/openshift/console/pull/3056/files#diff-626eda35061076f284f6cec7908679f4R37).

```suggestion
  return sections.map((section) => ({
     Component: getResourceTabSectionComp(section),
     key: section.properties.key
  }));
```

File: frontend/packages/dev-console/src/components/pipelines/pipeline-overview/PipelineOverview.tsx [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+export const PipelineRunItem: React.FC<PipelineRunItemProps> = ({ pipelineRun }) => {
+  const {
+    metadata: { name, namespace, creationTimestamp },
+    status: { completionTimestamp, startTimestamp },
+  } = pipelineRun;
+  const path = resourcePath(referenceForModel(PipelineRunModel), name, namespace);
+  const lastUpdated = completionTimestamp || startTimestamp || creationTimestamp;
+  return (
+    <li className="list-group-item">
+      <Grid gutter="md">
```
Comment (andrewballantyne) [inline frontend/packages/dev-console/src/components/pipelines/pipeline-overview/PipelineOverview.tsx]: I feel this `Grid` is unbalanced as well, but it aligns with `Pods` so we'll have to keep this. I think we could definitely improve the usage of space here, there is unnecessary whitespace gaps.

File: frontend/packages/dev-console/src/components/pipelines/pipeline-overview/TriggerLastRunButton.tsx [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+
+const TriggerLastRunButton: React.FC<TriggerLastRunButtonProps> = ({ pipeline, impersonate }) => {
+  const latestRun = _.get(pipeline, ['latestRun'], null);
+  const { label, callback, accessReview } = rerunPipeline(PipelineModel, pipeline, latestRun);
+  const isAllowed = useAccessReview(accessReview, impersonate);
+  return isAllowed ? (
+    <Button variant="secondary" onClick={callback}>
+      {label}
+    </Button>
+  ) : null;
```
Comment (andrewballantyne) [inline frontend/packages/dev-console/src/components/pipelines/pipeline-overview/TriggerLastRunButton.tsx]: You almost never need to condition a render between an Element and `null` in React. Falsey values are not rendered. Recommend just simplifying this to:

```
  return (
    isAllowed && (
      <Button variant="secondary" onClick={callback}>
        {label}
      </Button>
    )
  );
```

Alternatively you can just return `null` if not allowed and then after that return the `Button`. Both are fine with me.

File: frontend/packages/dev-console/src/components/pipelines/pipeline-overview/PipelineOverview.tsx [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+  ResourceLink,
+  resourcePath,
+} from '@console/internal/components/utils';
+import { PipelineRunModel, PipelineModel } from '@console/dev-console/src/models';
+import { Link } from 'react-router-dom';
+import { fromNow } from '@console/internal/components/utils/datetime';
+import { pipelineRunFilterReducer } from '../../../utils/pipeline-filter-reducer';
+import TriggerLastRunButton from './TriggerLastRunButton';
+
+const noOfPipelineRunsToshow = 3;
```
Comment (andrewballantyne) [inline frontend/packages/dev-console/src/components/pipelines/pipeline-overview/PipelineOverview.tsx]: Optional change...
```suggestion
const MAX_VISIBLE = 3;
```
Could go with a shorter name so it doesn't read like a phrase.

File: frontend/packages/dev-console/src/components/pipelines/pipeline-overview/PipelineOverview.tsx [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+    </li>
+  );
+};
+
+type PipelinesOverviewProps = {
+  item: OverviewItem & {
+    pipeline: K8sResourceKind[];
+    pipelineRuns: K8sResourceKind[];
+  };
+};
```
Comment (andrewballantyne) [inline frontend/packages/dev-console/src/components/pipelines/pipeline-overview/PipelineOverview.tsx]: Just an FYI, we are going to have overlap / conflict here. My PR for the tooltips has now been merged, so you'll probably want to rebase and adopt the changes I made to simplify things.

- Renamed `pipeline` to `pipelines` as it's an array - [REF](https://github.com/openshift/console/pull/3095/files#diff-bbb2e3628eb7d5b19c6b638c0968dc9dR31)
- Made this type in my PR - probably needs a rename to make use for both of us though - [REF](https://github.com/openshift/console/pull/3095/files#diff-5f5edb39a5aa09e8825496aa61c85f76R65)

File: frontend/packages/dev-console/src/components/pipelines/pipeline-overview/TriggerLastRunButton.tsx [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+import { impersonateStateToProps } from '@console/internal/reducers/ui';
+import { K8sResourceKind } from '@console/internal/module/k8s';
+import { useAccessReview } from '@console/internal/components/utils';
+import { Button } from 'patternfly-react';
+import { rerunPipeline } from '../../../utils/pipeline-actions';
+import { PipelineModel } from '../../../models';
+
+type TriggerLastRunButtonProps = {
+  pipeline: K8sResourceKind;
+  impersonate?;
```
Comment (andrewballantyne) [inline frontend/packages/dev-console/src/components/pipelines/pipeline-overview/TriggerLastRunButton.tsx]: I know it's not really typed anywhere - but we should definitely look to avoid untyped props. Could you check to see if you can type this?

File: frontend/packages/dev-console/src/components/pipelines/pipeline-overview/PipelineOverview.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (sahil143) [reply frontend/packages/dev-console/src/components/pipelines/pipeline-overview/PipelineOverview.tsx]: inline-block is used to align it with `View all ()`

File: frontend/public/components/overview/resource-overview-page.tsx (hunk 4 above)
Comment (sahil143) [reply frontend/public/components/overview/resource-overview-page.tsx]: Yes, Not sure what is happening here, looking into this

File: frontend/packages/dev-console/src/components/pipelines/pipeline-overview/PipelineOverview.tsx [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+        <Link
+          className="sidebar__section-view-all"
+          to={`${resourcePath(referenceForModel(PipelineModel), name, namespace)}/Runs`}
+        >
+          {`View All (${pipelineRuns.length})`}
+        </Link>
+      )}
+      <ListGroup componentClass="ul">
+        <li className="list-group-item pipeline-overview">
+          <Flex breakpointMods={[{ modifier: 'justify-content-space-between' }] as any}>
```
Comment (sahil143) [inline frontend/packages/dev-console/src/components/pipelines/pipeline-overview/PipelineOverview.tsx]: any is used here to avoid TypeError because breakpoint in `breakpointMods` is not optional and their is no need to use breakpoint with `justify-content-space-between`. Also, Typing for `breakpointMods` suggests `justify-content-flex-space-between` which is invalid and undefined. Correct value is `justify-content-space-between`
https://patternfly-react.surge.sh/patternfly-4/documentation/react/layouts/flex#justify-content-space-between

File: frontend/packages/dev-console/src/components/pipelines/pipeline-overview/PipelineOverview.tsx (hunk 1 above)
Comment (sahil143) [reply frontend/packages/dev-console/src/components/pipelines/pipeline-overview/PipelineOverview.tsx]: fixed in https://github.com/openshift/console/pull/3056/commits/8e27cbfc8c376c202bd2471c788bf0401b8dcb9d

File: frontend/packages/dev-console/src/components/pipelines/pipeline-overview/PipelineOverview.tsx (hunk 2 above)
Comment (sahil143) [reply frontend/packages/dev-console/src/components/pipelines/pipeline-overview/PipelineOverview.tsx]: fixed in https://github.com/openshift/console/pull/3056/commits/8e27cbfc8c376c202bd2471c788bf0401b8dcb9d

File: frontend/packages/dev-console/src/components/pipelines/pipeline-overview/PipelineOverview.tsx (hunk 3 above)
Comment (sahil143) [reply frontend/packages/dev-console/src/components/pipelines/pipeline-overview/PipelineOverview.tsx]: fixed in https://github.com/openshift/console/pull/3056/commits/8e27cbfc8c376c202bd2471c788bf0401b8dcb9d

File: frontend/packages/dev-console/src/components/pipelines/pipeline-overview/PipelineOverview.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (sahil143) [reply frontend/packages/dev-console/src/components/pipelines/pipeline-overview/PipelineOverview.tsx]: fixed plural label in https://github.com/openshift/console/pull/3056/commits/8e27cbfc8c376c202bd2471c788bf0401b8dcb9d

File: frontend/packages/dev-console/src/components/pipelines/pipeline-overview/PipelineOverview.tsx (hunk 6 above)
Comment (sahil143) [reply frontend/packages/dev-console/src/components/pipelines/pipeline-overview/PipelineOverview.tsx]: fixed in https://github.com/openshift/console/pull/3056/commits/8e27cbfc8c376c202bd2471c788bf0401b8dcb9d

File: frontend/packages/dev-console/src/components/pipelines/pipeline-overview/PipelineOverview.tsx (hunk 7 above)
Comment (sahil143) [reply frontend/packages/dev-console/src/components/pipelines/pipeline-overview/PipelineOverview.tsx]: fixed in https://github.com/openshift/console/pull/3056/commits/8e27cbfc8c376c202bd2471c788bf0401b8dcb9d

File: frontend/public/components/overview/resource-overview-page.tsx (hunk 8 above)
Comment (sahil143) [reply frontend/public/components/overview/resource-overview-page.tsx]: fixed in https://github.com/openshift/console/pull/3056/commits/8e27cbfc8c376c202bd2471c788bf0401b8dcb9d

File: frontend/packages/dev-console/src/components/pipelines/pipeline-overview/PipelineOverview.tsx (hunk 12 above)
Comment (sahil143) [reply frontend/packages/dev-console/src/components/pipelines/pipeline-overview/PipelineOverview.tsx]: fixed in https://github.com/openshift/console/pull/3056/commits/8e27cbfc8c376c202bd2471c788bf0401b8dcb9d

File: frontend/packages/dev-console/src/components/pipelines/pipeline-overview/PipelineOverview.tsx (hunk 11 above)
Comment (sahil143) [reply frontend/packages/dev-console/src/components/pipelines/pipeline-overview/PipelineOverview.tsx]: fixed in https://github.com/openshift/console/pull/3056/commits/8e27cbfc8c376c202bd2471c788bf0401b8dcb9d

File: frontend/packages/dev-console/src/components/pipelines/pipeline-overview/TriggerLastRunButton.tsx (hunk 10 above)
Comment (sahil143) [reply frontend/packages/dev-console/src/components/pipelines/pipeline-overview/TriggerLastRunButton.tsx]: fixed in https://github.com/openshift/console/pull/3056/commits/8e27cbfc8c376c202bd2471c788bf0401b8dcb9d

File: frontend/public/components/overview/resource-overview-page.tsx (hunk 8 above)
Comment (sahil143) [reply frontend/public/components/overview/resource-overview-page.tsx]: fixed in https://github.com/openshift/console/pull/3056/commits/8e27cbfc8c376c202bd2471c788bf0401b8dcb9d

File: frontend/packages/dev-console/src/components/pipelines/pipeline-overview/PipelineOverview.tsx [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+  },
+}) => {
+  const {
+    metadata: { name, namespace },
+  } = pipeline;
+  return (
+    <>
+      <SidebarSectionHeading
+        text={PipelineRunModel.labelPlural}
+        style={{ display: 'inline-block' }}
```
Comment (andrewballantyne) [inline frontend/packages/dev-console/src/components/pipelines/pipeline-overview/PipelineOverview.tsx]: Inline styles should be avoided when possible.

In this case, Patternfly has an odd implementation of the flexbox, which gives us a simple solution:
```
      <Flex>
        <FlexItem>
          <SidebarSectionHeading text={PipelineRunModel.labelPlural}/>
        </FlexItem>
        <FlexItem>
          {pipelineRuns.length > MAX_VISIBLE && (
            <Link
              className="sidebar__section-view-all"
              to={`${resourcePath(referenceForModel(PipelineModel), name, namespace)}/Runs`}
            >
              {`View All (${pipelineRuns.length})`}
            </Link>
          )}
        </FlexItem>
      </Flex>
```

File: frontend/packages/dev-console/src/components/pipelines/pipeline-overview/PipelineOverview.tsx [hunk 16]
```diff
@@ -0,0 +1,70 @@
...
+            <FlexItem>
+              <ResourceLink
+                inline
+                kind={referenceForModel(PipelineModel)}
+                name={name}
+                namespace={namespace}
+              />
+            </FlexItem>
+            <FlexItem>
+              <TriggerLastRunButton pipeline={pipeline} />
```
[the remark this answers is not in the document]
Comment (andrewballantyne) [reply frontend/packages/dev-console/src/components/pipelines/pipeline-overview/PipelineOverview.tsx]: And when we have never ran, we don't show "last run":

image

Can't start a `last run` when there is no `last run`.

File: frontend/public/components/overview/resource-overview-page.tsx [hunk 17]
```diff
@@ -9,19 +9,41 @@ import { NetworkingOverview } from './networking-overview';
...
+  const pluginComponents = getPluginTabSectionResource(item);
+  return (
+    <div className="overview__sidebar-pane-body">
+      <PodsOverview pods={pods} obj={obj} />
+      <BuildOverview buildConfigs={buildConfigs} />
+      {pluginComponents.map(({ Component, key }) => (
+        <Component key={key} item={item} />
+      ))}
+      <NetworkingOverview services={services} routes={routes} />
+    </div>
```
[the remark this answers is not in the document]
Comment (sahil143) [reply frontend/public/components/overview/resource-overview-page.tsx]: fixed in https://github.com/openshift/console/pull/3056/commits/ed5ac16c643177366ba3b34dcfa8141b3d172438

File: frontend/packages/dev-console/src/components/pipelines/pipeline-overview/TriggerLastRunButton.tsx [hunk 18]
```diff
@@ -0,0 +1,29 @@
...
+};
+
+const TriggerLastRunButton: React.FC<TriggerLastRunButtonProps> = ({ pipeline, impersonate }) => {
+  const latestRun = _.get(pipeline, ['latestRun'], null);
+  const { label, callback, accessReview } = rerunPipeline(PipelineModel, pipeline, latestRun);
+  const isAllowed = useAccessReview(accessReview, impersonate);
+  return (
+    isAllowed && (
+      <Button variant="secondary" onClick={callback}>
```
[the remark this answers is not in the document]
Comment (sahil143) [reply frontend/packages/dev-console/src/components/pipelines/pipeline-overview/TriggerLastRunButton.tsx]: https://github.com/openshift/console/pull/3056/commits/ed5ac16c643177366ba3b34dcfa8141b3d172438

File: frontend/packages/dev-console/src/components/pipelines/pipeline-overview/PipelineOverview.tsx [hunk 19]
```diff
@@ -0,0 +1,70 @@
...
+
+const MAX_VISIBLE = 3;
+
+type PipelinesOverviewProps = {
+  item: TopologyOverviewItem;
+};
+
+const PipelinesOverview: React.FC<PipelinesOverviewProps> = ({
+  item: {
+    pipelines: [pipeline],
```
[the remark this answers is not in the document]
Comment (sahil143) [reply frontend/packages/dev-console/src/components/pipelines/pipeline-overview/PipelineOverview.tsx]: fixed in https://github.com/openshift/console/pull/3056/commits/ed5ac16c643177366ba3b34dcfa8141b3d172438

File: frontend/packages/dev-console/src/components/pipelines/pipeline-overview/PipelineOverview.tsx (hunk 15 above)
Comment (sahil143) [reply frontend/packages/dev-console/src/components/pipelines/pipeline-overview/PipelineOverview.tsx]: fixed in https://github.com/openshift/console/pull/3056/commits/ed5ac16c643177366ba3b34dcfa8141b3d172438. Just checked that SidebarSectionHeading takes children prop also.

File: frontend/packages/dev-console/src/components/pipelines/pipeline-overview/PipelineOverview.tsx (hunk 1 above)
Comment (andrewballantyne) [reply frontend/packages/dev-console/src/components/pipelines/pipeline-overview/PipelineOverview.tsx]: I missed it in my second pass - it should be a default export as well.

File: frontend/packages/dev-console/src/components/pipelines/pipeline-overview/PipelineOverview.tsx (hunk 14 above)
Comment (andrewballantyne) [reply frontend/packages/dev-console/src/components/pipelines/pipeline-overview/PipelineOverview.tsx]: Probably needs an issue over here: https://github.com/patternfly/patternfly-react/issues

We should look to report bugs when we notice them, especially from Patternfly since we use it quite often.

Even when deep-diving into their types I have found:
```
export declare type FlexBreakpointMod = {
  ...
  modifier: 'flex' | ... | 'justify-content-flex-space-between' | ...
  ...
}
```
Which still gives an unfortunate error:
```
ERROR in /Users/andrewballantyne/code/others/console/frontend/packages/dev-console/src/components/pipelines/pipeline-overview/PipelineOverview.tsx(45,36):
TS2322: Type 'string' is not assignable to type '"column" | "flex" | "inline-flex" | "column-reverse" | "row" | "row-reverse" | "nowrap" | "wrap" | "wrap-reverse" | "full-width" | "spacer-none" | "spacer-xs" | "spacer-sm" | ... 44 more ... | "align-content-space-around"'.
```
Not sure what it means by "string" - as they are all strings :/

Their docs don't match up, nor does following the type. Something is up, definitely recommend you log an issue over on their GitHub @sahil143 - let me know if you need anything from me around this.

Comment (sahil143) [reply frontend/packages/dev-console/src/components/pipelines/pipeline-overview/PipelineOverview.tsx]: Yes, I will be logging an issue for this. Docs says to use `justify-content-space-between` whereas typing for FlexBreakpointMod suggests `'justify-content-flex-space-between'` which is really weird.
> Not sure what it means by "string" - as they are all strings :/

You might be getting this because you didn't define breakpoint second prop of `FlexBreakpoint Mod`.
Also, the breakpoint prop should be optional here.

Comment (andrewballantyne) [reply frontend/packages/dev-console/src/components/pipelines/pipeline-overview/PipelineOverview.tsx]: Nope. I definitely get two different issues with `breakpoint` and `breakpointMods`:

```
ERROR in /.../dev-console/src/components/pipelines/pipeline-overview/PipelineOverview.tsx(46,36):
TS2322: Type 'string' is not assignable to type '"column" | "flex" | "inline-flex" | "column-reverse" | "row" | "row-reverse" | "nowrap" | "wrap" | "wrap-reverse" | "full-width" | "spacer-none" | "spacer-xs" | "spacer-sm" | ... 44 more ... | "align-content-space-around"'.
ERROR in /.../dev-console/src/components/pipelines/pipeline-overview/PipelineOverview.tsx(46,79):
TS2322: Type 'string' is not assignable to type '"sm" | "md" | "lg" | "xl" | "2xl"'.
Version: typescript 3.6.3
```
```
<Flex breakpointMods={[{ modifier: 'justify-content-space-between', breakpoint: 'sm' }]}>
```
Not sure what is up - I'd expect this to work.

Comment (sahil143) [reply frontend/packages/dev-console/src/components/pipelines/pipeline-overview/PipelineOverview.tsx]: Yes this will not work because it's expecting `'justify-content-flex-space-between'` instead of `'justify-content-space-between'`
```
<Flex breakpointMods={[{ modifier: 'justify-content-flex-space-between', breakpoint: 'sm' }]}>
```

Comment (andrewballantyne) [reply frontend/packages/dev-console/src/components/pipelines/pipeline-overview/PipelineOverview.tsx]:
```suggestion
          <Flex breakpointMods={[{ modifier: 'justify-content-flex-space-between' }] as any}>
```
:)

Comment (sahil143) [reply frontend/packages/dev-console/src/components/pipelines/pipeline-overview/PipelineOverview.tsx]: `'justify-content-flex-space-between'` only resolves the TypeError, not the actual problem. It should be `justify-content-space-between` in type definitions. `'justify-content-flex-space-between'` is wrong and not defined in patternfly-react styles. It just returns undefined.

File: frontend/packages/dev-console/src/components/pipelines/pipeline-overview/PipelineOverview.tsx (hunk 16 above)
[the remark this answers is not in the document]
Comment (sahil143) [reply frontend/packages/dev-console/src/components/pipelines/pipeline-overview/PipelineOverview.tsx]: Disabled the Start Last run button when there are no pipelineruns https://github.com/openshift/console/pull/3056/commits/a38f07aebad6e5cd815adc6423da819833af7f40

File: frontend/packages/dev-console/src/components/pipelines/pipeline-overview/PipelineOverview.tsx (hunk 14 above)
Comment (andrewballantyne) [reply frontend/packages/dev-console/src/components/pipelines/pipeline-overview/PipelineOverview.tsx]: Please do log an issue with Patternfly @sahil143

File: frontend/packages/dev-console/src/components/pipelines/pipeline-overview/TriggerLastRunButton.tsx [hunk 20]
```diff
@@ -0,0 +1,34 @@
...
+import { impersonateStateToProps } from '@console/internal/reducers/ui';
+import { K8sResourceKind } from '@console/internal/module/k8s';
+import { useAccessReview } from '@console/internal/components/utils';
+import { Button } from 'patternfly-react';
+import { rerunPipeline } from '../../../utils/pipeline-actions';
+import { PipelineModel } from '../../../models';
+
+type TriggerLastRunButtonProps = {
+  pipeline: K8sResourceKind;
+  disable: boolean;
```
Comment (andrewballantyne) [inline frontend/packages/dev-console/src/components/pipelines/pipeline-overview/TriggerLastRunButton.tsx]: This should probably be called `disabled` and should be optional for the purity of the component.

That way the interface is similar to that a button:
```
<TriggerLastRunButton disabled />
<Button disabled />
```

File: frontend/public/components/overview/resource-overview-page.tsx (hunk 4 above)
Comment (sahil143) [reply frontend/public/components/overview/resource-overview-page.tsx]: Jira Issue: https://jira.coreos.com/browse/ODC-2122

File: frontend/packages/dev-console/src/components/pipelines/pipeline-overview/PipelineOverview.tsx (hunk 14 above)
Comment (sahil143) [reply frontend/packages/dev-console/src/components/pipelines/pipeline-overview/PipelineOverview.tsx]: patternfly issue: https://github.com/patternfly/patternfly-react/issues/3244

**Comment by andrewballantyne** [comment]
> /lgtm

@rohitkrai03 could you do an approval?
